### PR TITLE
[Docs] Add examples for new object-based c10d APIs

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43932 [Docs] Add examples for new object-based c10d APIs**
* #48909 Add object-based collective APIs to public docs

Adds some basic examples to the documentation for each of the newly added
object-based collectives.

Differential Revision: [D23441838](https://our.internmc.facebook.com/intern/diff/D23441838/)

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -1393,6 +1393,15 @@ def all_gather_object(object_list, obj, group=group.WORLD):
...
         function with data you trust.
+
+    Example::
+        >>> # Note: Process group initialization omitted on each rank.
+        >>> import torch.distributed as dist
+        >>> gather_objects = ["foo", 12, {1: 2}] # any picklable object
+        >>> output = [None for _ in gather_objects]
+        >>> dist.all_gather_object(output, gather_objects[dist.get_rank()])
+        >>> output
+        ['foo', 12, {1: 2}]
```
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py:1404]: shall we verify if this line is rendered correctly?

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -1467,6 +1476,20 @@ def gather_object(obj, object_gather_list=None, dst=0, group=group.WORLD):
...
+    Example::
+        >>> # Note: Process group initialization omitted on each rank.
+        >>> import torch.distributed as dist
+        >>> gather_objects = ["foo", 12, {1: 2}] # any picklable object
+        >>> output = [None for _ in gather_objects]
+        >>> dist.gather_object(
+                gather_objects[dist.get_rank()],
+                output if dist.get_rank() == 0 else None,
+                dst=0
+            )
```
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py:1489]: shall we also verify if these lines are rendered correctly?

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -1556,6 +1579,17 @@ def broadcast_object_list(object_list, src, group=group.WORLD):
...
+
+    Example::
+        >>> # Note: Process group initialization omitted on each rank.
+        >>> import torch.distributed as dist
+        >>> if dist.get_rank() == 0:
+        >>>     objects = ["foo", 12, {1: 2}] # any picklable object
+        >>> else:
+        >>>     objects = [None, None, None]
+        >>> dist.broadcast_object_list(objects, src=0)
+        >>> broadcast_objects
```
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py:1591]: this var is not defined?

File: torch/distributed/distributed_c10d.py [hunk 4]
```diff
@@ -1393,6 +1393,15 @@ def all_gather_object(object_list, obj, group=group.WORLD):
         known to be insecure. It is possible to construct malicious pickle data
         which will execute arbitrary code during unpickling. Only call this
         function with data you trust.
+
+    Example::
+        >>> # Note: Process group initialization omitted on each rank.
+        >>> import torch.distributed as dist
+        >>> gather_objects = ["foo", 12, {1: 2}] # any picklable object
```
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py:1400]: this assumes `world_size==3`?

File: torch/distributed/distributed_c10d.py [hunk 5]
```diff
@@ -1634,6 +1668,19 @@ def scatter_object_list(
...
         data which will execute arbitrary code during unpickling. Only call this
         function with data you trust.
+
+    Example::
+        >>> # Note: Process group initialization omitted on each rank.
+        >>> import torch.distributed as dist
+        >>> if dist.get_rank() == 0:
+        >>>     objects = ["foo", 12, {1: 2}] # any picklable object
+        >>> else:
+        >>>     objects = [None, None, None]
```
Comment (mrshenli) [inline torch/distributed/distributed_c10d.py:1678]: does this have to be 3 Nones?

**Review by mrshenli — APPROVED** [review]
LGTM! Thanks for adding this! Left some minor comments mainly about rendering and clarity.

File: torch/distributed/distributed_c10d.py (hunk 5 above)
Comment (rohan-varma) [reply torch/distributed/distributed_c10d.py:1678]: It's mostly to point out that whatever is contained in `objects` on non-src ranks, it doesn't matter, we will only look at the `objects` from the src rank. The 2nd arg can be anything for non-src ranks, it is not considered.

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (rohan-varma) [reply torch/distributed/distributed_c10d.py:1591]: Good catch, it should be just `objects`.

File: torch/distributed/distributed_c10d.py (hunk 4 above)
Comment (rohan-varma) [reply torch/distributed/distributed_c10d.py:1400]: Right, I guess we could just add this as a comment to keep the example clear and concise.

**Comment by rohan-varma** [comment]
@mrshenli Looks like the rendering is as expected after building the docs locally, see for example: 

all_gather_ss

**Pull request closed without merging**